backend/api/src/api/v1/gql/chat.rs

```diff
@@ -0,0 +1,131 @@
+            .clone();
+
+        let request_context = ctx
+            .data::<Arc<GqlContext>>()
+            .expect("failed to get request context");
+
+        if content.len() > 500 {
+            return Err(GqlError::InvalidInput.with_message("Message too long"));
+        }
+
+        let session = request_context.get_session(&global).await?;
+
+        let user_id = session
+            .as_ref()
+            .ok_or_else(|| GqlError::InvalidInput.with_message("You need to be logged in"))?
+            .user_id;
+
+        let user = sqlx::query!("SELECT username FROM users WHERE id = $1", user_id,)
+            .map(|row| user::Model {
+                username: row.username,
+                ..Default::default()
+            })
+            .fetch_optional(&*global.db)
+            .await
+            .map_err(|_| GqlError::InvalidInput.with_message("Failed to fetch user"))?
+            .ok_or_else(|| GqlError::InvalidInput.with_message("User not found"))?;
+
+        let _chat = sqlx::query!("SELECT id FROM chat_rooms WHERE id = $1", chat_id,)
+            .map(|row| chat_room::Model {
+                id: row.id,
+                ..Default::default()
+            })
+            .fetch_optional(&*global.db)
+            .await
+            .map_err(|_| GqlError::InvalidInput.with_message("Failed to fetch chat room"))?
+            .ok_or_else(|| GqlError::InvalidInput.with_message("Chat not found"))?;
+
+        let message_to_send = serde_json::json!({ "chat_id": chat_id, "username": user.username, "content": content, "message_type": "message" }).to_string();
+        match global
+            .redis_pool
+            .publish::<String, String, String>(
+                "chat:".to_string() + &chat_id.to_string(),
+                message_to_send.to_string(),
+            )
+            .await
+        {
+            Ok(_) => {}
+            Err(_) => {
+                return Err(GqlError::InternalServerError.with_message("Failed to publish message"));
+            }
+        };
+
+        let _ = sqlx::query!(
+            "INSERT INTO chat_messages (chat_room_id, author_id, message, created_at) VALUES ($1, $2, $3, $4)",
+            chat_id,
+            user_id,
+            content,
+            Utc::now()
+        )
+        .execute(&*global.db)
+        .await;
+
+        Ok(message_to_send)
+    }
+}
+
+#[Subscription]
+impl ChatSubscription {
+    // Listen to new messages in chat.
+    pub async fn new_message<'ctx>(
+        &self,
+        ctx: &Context<'_>,
+        #[graphql(desc = "Chat to subscribe to.")] chat_id: i64,
+    ) -> impl Stream<Item = Message> {
+        let global = ctx
+            .data::<Arc<GlobalState>>()
+            .expect("failed to get global state")
+            .clone();
+
+        let mut message_stream = global.redis_sub_client.on_message();
+        let _ = global
+            .redis_sub_client
+            .subscribe("chat:".to_string() + &chat_id.to_string())
+            .await;
```

Comment:
Not sure of this approach, it seems that you have a single global broadcast channel?
Perhaps we need to change how this works.

Reply:
Do you have any idea?